source/plugin/modularization.rst

```diff
@@ -0,0 +1,61 @@
+===================
+Code Modularization
+===================
+
+When writing Sponge plugins, you may find that you need to modularize your code. Maybe you want to split your core into
+another project so that others can make use of it, or maybe you want to utilize a service another plugin provides. With
+Sponge, gone are the days of shading entire plugins into your plugin.
+
+Problem 1: Providing Good Service
+=================================
+
+Let's say you want to provide some awesome service for ``WarpService``. You want to make ``AwesomeWarpService`` override
+some other plugin's ``CoreWarpService`` with your own.
+
+But when you go to use ``Sponge.getServiceManager().provide(WarpService.class)``, you are shocked to find that referencing
```

Comment:
No, plugins are kept on separate environments unless explicitly depended on.


Get Outlook for Android






On Sat, Jul 16, 2016 at 9:21 PM -0500, "Ryan Leach" <notifications@github.com> wrote:












In source/plugin/modularization.rst:
> @@ -0,0 +1,61 @@
> +===================
> +Code Modularization
> +===================
> +
> +When writing Sponge plugins, you may find that you need to modularize your code. Maybe you want to split your core into
> +another project so that others can make use of it, or maybe you want to utilize a service another plugin provides. With
> +Sponge, gone are the days of shading entire plugins into your plugin.
> +
> +Problem 1: Providing Good Service
> +=================================
> +
> +Let's say you want to provide some awesome service for ``WarpService``. You want to make ``AwesomeWarpService`` override
> +some other plugin's ``CoreWarpService`` with your own.
> +
> +But when you go to use ``Sponge.getServiceManager().provide(WarpService.class)``, you are shocked to find that referencing



When? I would have thought that unless this is a static initializer or constructor that the other plugins would have been classloaded by the time you actually access the service manager.



—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.